### PR TITLE
Supporting bandwidth / bandScale

### LINKED ISSUE #26: Allow configuration of bar series offset
The current bar series (and all other series with some notion of width) is implemented to render around the centre point of each location across the cross-axis. In other words, values on the cross-axis represent the centres of each bar.

This works well for rendering bar series where the cross axis is continuous (e.g. linear or time), however the [d3 ordinal scale](https://github.com/d3/d3-scale#ordinal-scales) are more suited to series where the cross axis values represent the start of each bar.

As a result the bar series do not play well with d3 band scales.

In order to solve this, the code which determines the alignment of the underlying d3fc-shape should be configurable:

https://github.com/d3fc/d3fc-series/blob/master/src/svg/bar.js#L41

### PR DESCRIPTION
**Pull request by ColinEberhardt**
fixes #26 #44

This is a quick implementation to try out the concept for bar series. Two new properties are added, `bandwidth` (which mirrors the `d3.scaleBand` [API](https://github.com/d3/d3-scale#band_bandwidth)), and `align`. As a result, the bar series now pays nicely with band scales:

```
var xScale = d3.scaleBand()
    .domain(data.map(d => d.x))
    .paddingInner(0.2)
    .paddingOuter(0.1)
    .rangeRound([0, width]);

var svgBar = fc.seriesSvgBar()
    .xScale(xScale)
    .yScale(yScale)
    .align('right')
    .bandwidth(xScale.bandwidth())
    .crossValue(function(d) { return d.x; })
    .mainValue(function(d) { return d.y; });
```

For backwards compatibility, if bandwidth is not set the series adopts the current behaviour.

Note, the `align` property is required because for non-band scales you'd expect the bar to render with centre alignment (lines scales are similar to point scales), whereas for the band scale, the convention defined by D3 require a right-aligned bar.

Open questions:
 - should this be extended to all series types that have a width? i.e. box plot, candlestick, ohlc.
 - It is natural to plot bars on a bandScale, but not candlestick, perhaps the current API should not be considered deprecated.
 - With a cartesian chart, when do we get the opportunity to obtain the bandwidth from the scale?, i.e. `barSeries.bandwidth(xScale.bandwidth())` - is this going to have to be a 'magic' feature of the cartesian component?

### CODE REVIEW

**Comment by ColinEberhardt** [comment]
Thoughts ...

- xyBase should detect bandwidth on the scale and use that as a default
- move fractionalBarWidth to be a wrapper around the series
- change barWidth to bandwidth and have a default of some sensible scalar, e.g. 5

**Comment by ColinEberhardt** [comment]
@chrisprice just quickly put together an `autoBandwidth` wrapper component whilst trundling along on the metro 🚇 - it works really well. Minimal changes to all of the other series types. The `autoBandwidth` works with both SVG & Canvas, and also includes the little bit of magic around bandScale where we detect and use the `bandwidth` function. It was a better fit there than `xyBase`.

File: src/autoBandwidth.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import { select, selection } from 'd3-selection';
+import { rebindAll } from 'd3fc-rebind';
+import { pairs, min } from 'd3-array';
+
+export default (adaptee) => {
+
+    let widthFraction = 0.75;
+
+    // the barWidth property of the various series takes a function which, when given an
```
Comment (chrisprice) [inline src/autoBandwidth.js]: barWidth -> bandwidth

File: src/autoBandwidth.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                    .map(adaptee.crossValue())
+                    .map(crossScale);
+
+                const width = computeBarWidth(crossValuePoints);
+
+                adaptee.bandwidth(width);
+            }
+        };
+
+        if (arg instanceof selection) {
```
Comment (chrisprice) [inline src/autoBandwidth.js]: Not sure when it wouldn't be a selection?

File: src/autoBandwidth.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            });
+        } else {
+            computeWidth(arg);
+            adaptee(arg);
+        }
+    };
+
+    rebindAll(autoBandwidth, adaptee);
+
+    autoBandwidth.widthFraction = (...args) => {
```
Comment (chrisprice) [inline src/autoBandwidth.js]: bandwidth?

File: src/autoBandwidth.js (hunk 2 above)
Comment (ColinEberhardt) [inline src/autoBandwidth.js]: When used to adapt a canvas series component

**Comment by ColinEberhardt** [comment]
Thanks 👍

**Comment by ColinEberhardt** [comment]
@chrisprice I've finished the implementation of bandwidth, which you can see in this commit https://github.com/d3fc/d3fc-series/pull/46/commits/28c79296e2042fc65a6de222c41a69b698ad6de5 - as discussed, this works very well.

The interesting part is the next commit which adds support for variable bandwidths with grouped series https://github.com/d3fc/d3fc-series/pull/46/commits/b508b017dab6ae9e46bda8b5e7466e459c9dafa2 

This allows funky things like this:

screen shot 2017-04-05 at 21 50 53

It works by creating an offset scale that varies based on index. This requires the slightly controversial change where scales are invoked with both the domain value and index:
https://github.com/d3fc/d3fc-series/pull/46/commits/b508b017dab6ae9e46bda8b5e7466e459c9dafa2#diff-69ce48de4a59dbea5fd813281f833614R27

File: README.md [hunk 4]
```diff
@@ -267,15 +274,44 @@ In this case the context is also propagated from the multi series to the childre
...
+    .align('left')
+    .crossValue(function(d) { return d.x; })
+    .mainValue(function(d) { return d.y; });
+```
+
+Notice in the above example that the `align` property of the bar is set to `left`, which reflects the band scale coordinate system.
+
+**NOTE:** The d3 point scale is a special cased band scale that has a zero bandwidth. As a result, if you use the `fc.autoBandwidth` component in conjunction with a point scale, the series will also have a bandwidth of zero!
+
+When used in conjunction with a linear scale, the `fc.autoBandwidth` component compute the bar width based on the smallest distance between consecutive datapoints:
```
Comment (chrisprice) [inline README.md]: component compute_s_

File: README.md [hunk 5]
```diff
@@ -267,15 +274,44 @@ In this case the context is also propagated from the multi series to the childre
...
+    .domain([0, 10])
+    .range([0, width]);
+
+var svgBar = fc.autoBandwidth(fc.seriesSvgBar())
+    .crossValue(function(d) { return d.x; })
+    .mainValue(function(d) { return d.y; })
+    .widthFraction(0.5);
+```
+
+The `fc.autoBandwidth` component, when adapting a series, adds a `widthFraction` property which determines the fraction of this distance that if used to set the bandwidth.
```
Comment (chrisprice) [inline README.md]: ~~if~~ is

File: examples/grouped.js [hunk 6]
```diff
@@ -77,8 +78,47 @@ var groupedBar = fc.seriesSvgGrouped(groupedSeries)
...
+    .padding(0.5)
+    .range([0, width]);
+
+groupedBar.xScale(x3)
+  .crossValue(function(d) { return d[0]; })
+  // centre align the bars around the points on the scale
+  .align('center')
+  // because point scales have a zero bandwidth, in this context we
+  // provide an explicit bandwidth and don't wrap the series in fc.autoBandwidth.
+  // this example also shows how bandwidth can very on a point-to-point basis
```
Comment (chrisprice) [inline examples/grouped.js]: v_~a~e_ry

**Pull request merged**